### ISSUE TITLE
CLS of 1.0 caused by "async" for https://cdn.ampproject.org/v0.js

### ISSUE THREAD

**Issue by Zerototh** [ampproject/amphtml#31003]
## What's the issue?

Async for https://cdn.ampproject.org/v0.js is causing 1 Cumulative Layout Shift that penalizes our amp-only site with a CLS of magnitude 1.0.

Removing async removes the CLS issue but the "async" for v0.js is mandatory for amp.

## How do we reproduce the issue?
Step 0. Make a boilerplate amp-html: https://amp.dev/boilerplate/
Step 1. Add some "lorem ipsum".
Step 2. Inspect performance with chrome, opera etc. CLS issue appears.
Step 3. Remove async from "<script async src="https://cdn.ampproject.org/v0.js"></script>"
Step 4. Inspect performance again and CLS issue is not there.
Step 5. Add async back on to double-check and the CLS is back.

The result is: 1 single CLS but with a magnitude of 1.0.

I've taken screen-shots of these steps.

##Browsers/Device?
Tested on chrome, opera, windows 10. On chrome I've inspected in both mobile- and desktop-mode.

## Which AMP version is affected?
Version 2010132225003

I got my first CLS error flag in Core Web Vitals at 20-10-2020 (2 weeks ago).

![cls issue 1](https://user-images.githubusercontent.com/73961755/98159200-720c9e00-1edc-11eb-8ff4-8e575f01ef14.png)
![cls issue 2](https://user-images.githubusercontent.com/73961755/98159202-72a53480-1edc-11eb-896c-d33811615f81.png)
![cls issue 3 async gone](https://user-images.githubusercontent.com/73961755/98159205-72a53480-1edc-11eb-8bf2-b44182619e09.png)
![cls issue 4 CLS gone](https://user-images.githubusercontent.com/73961755/98159206-733dcb00-1edc-11eb-8875-5a32abfb96d5.png)
![cls issue 5 Async back and CLS back](https://user-images.githubusercontent.com/73961755/98159207-733dcb00-1edc-11eb-8ae6-a40835288cf0.png)

**Comment by Sebschock** [comment]
I'm also experiencing this issue.

Performance with: `<script acync src="https://cdn.ampproject.org/lts/v0.js"></script>`
<img width="706" alt="Skærmbillede 2020-11-09 kl  13 59 34" src="https://user-images.githubusercontent.com/6289019/98544163-e3a96b00-2293-11eb-9b77-f235398a0ece.png">


Performance with: `<script src="https://cdn.ampproject.org/lts/v0.js"></script>`
<img width="708" alt="Skærmbillede 2020-11-09 kl  13 59 06" src="https://user-images.githubusercontent.com/6289019/98544216-f754d180-2293-11eb-905d-9df9465fb967.png">

**Comment by kristoferbaxter** [comment]
Triaging to @kevinkimball.

I'll also take a look when starting work today.

Theoretically, this makes little sense but we will root cause it.

**Comment by kristoferbaxter** [comment]
Given the example, I created a sample here to illustrate: https://lavender-jumpsuit.glitch.me/

1. As you can see this uses the 'async' attribute on `v0.js`.
<img width="579" alt="Screen Shot 2020-11-09 at 9 03 22 AM" src="https://user-images.githubusercontent.com/61764/98572411-7208f700-226a-11eb-9606-a6906099a490.png">

2. There is a Layout Shift once JavaScript loads (as expected).
<img width="724" alt="Screen Shot 2020-11-09 at 9 04 43 AM" src="https://user-images.githubusercontent.com/61764/98572578-a54b8600-226a-11eb-8dee-782cae9cc48b.png">

### This layout shift is from the execution of JavaScript on the document, removing the boilerplate.
There are several solutions to remove the layout shift with varying tradeoffs.
1. Use a sync script loading mechanism for `v0.js`. This is what the ticket is mentioning, however this puts the `v0.js` document as a sync script on the main loading timeline. 
2. **Preferred**: Use an AMP Optimizer to eliminate the boilerplate, removing the need for JS execution to do so.

When this approach is used (using an Optimizer or loading from Google's AMP Cache), the result is 0 CLS. An example is available here: https://lavender-jumpsuit.glitch.me/optimized.html

### So, in short... 
We recommend using an AMP Optimizer on canonical origins to eliminate the boilerplate since this will fix the issue and provides other origin based performance benefits in the long run.

**Comment by kristoferbaxter** [comment]
Marking this as closed, please comment if there are additional items to consider here and we can reopen the issue.

**Comment by kristoferbaxter** [comment]
Thank you for sharing your experience.

Can you link to the documents that are failing Core Web Vitals? We'd be happy to take a look.

If you'd prefer not to install an AMP Optimizer on your servers there are a few options:
1. Implement the functionality of an AMP Optimizer independently (the test suite is Open Source too).
2. Leverage the Open Source Docker implementation.
3. Contribute fixes to support other models like Cloudflare Worker support.

This specific issue (removing flashes of un-upgraded elements) is avoided by leveraging server side rendering (ie an AMP Optimizer) or by delaying content from appearing until it's settled (creates a CLS impact). This is why we recommend an Optimizer, which gives a better user experience.

**Comment by Zerototh** [comment]
I did as you suggested and installed AMP Optimizer. Our CLS is down to 0.008 which is way better than the 1.0 we had.

The best part is that we now have control over the cause of the CLS. We're now able to choose between real user experience and the metric CLS. Some things are worth their "CLS-weight". Something we didn't have control over before we installed the optimizer.

The performance was essentially the same as without it. With our site already using preload, srcsets, css minimization etc we didn't see much of a boost in performance.

Thank you for your help kristoferbaxter. Much appreciated.

**Comment by steffenweber** [comment]
The issue affects all of our AMP pages. Example article from today:

AMP: https://www.computerbase.de/2020-11/amazon-echo-dot-mit-uhr-test/?amp=1 (CLS = 1.0)
Web: https://www.computerbase.de/2020-11/amazon-echo-dot-mit-uhr-test/ (CLS = 0.0)

Another article that Search Console specifically complains about:

AMP: https://www.computerbase.de/thema/grafikkarte/rangliste/?amp=1 (CLS = 1.0)
Web: https://www.computerbase.de/thema/grafikkarte/rangliste/ (CLS = 0.0)

Tested using the [Web Vitals Chrome extension](https://chrome.google.com/webstore/detail/web-vitals/ahfhijdlegdabablpippeagghigmibma). I had to explicitly reload the page because otherwise the browser's cache makes the issue disappear.

I experimented with the [AMP Optimizer for PHP](https://github.com/ampproject/amp-toolbox-php) today. Out of the box it is prohibitively slow (it seems to be doing remote network requests on each invocation… :astonished:). Running only the "ServerSideRendering" transformation (which is enough to solve the CLS issue) makes it way faster but it'd still add 10–20 ms of PHP execution time to each page request. This would effectively double the PHP exectime for us.

**Comment by kristoferbaxter** [comment]
Unfortunately to remove the CLS concern on the origin AMP page, the documents need to be output with the SSR transforms applied. Otherwise the documents are either invalid, or will display rendering issues as the Custom Elements on the page are hydrated.

If the AMP Optimizer for PHP has a performance concern for your PHP environments, it might be prudent to work with the Open Source team to increase its efficiency.

**Comment by pmeenan** [comment]
Is this still an issue?  Chrome 89 shipped with some fixes for CLS measurement where it was reporting layout shifts for non-visible content (particularly the anti-flicker pattern of hiding with opacity: 0) and chrome 87 or 88 shipped with fixes for SPA page rehydration (though I think that was more LCP related).

I'm not seeing any layout shifts reported on the glitch.me test page: https://www.webpagetest.org/result/210308_AiCV_df4af78aea860c2e5bdf0089c972f378/1/details/#waterfall_view_step1

If it's still happening and the shifts aren't user visible then it might be another edge-case for the metric that needs to be looked at.  If the shifts are user visible then using an anti-flicker technique should be able to fix it and also not trigger a reported layout shift without having to install optimizer.
